stan/math/prim/fun/dot_product.hpp

```diff
@@ -17,11 +18,9 @@ namespace math {
  * @throw std::domain_error If the vectors are not the same
  * size or if they are both not vector dimensioned.
  */
-template <int R1, int C1, int R2, int C2>
-inline double dot_product(const Eigen::Matrix<double, R1, C1> &v1,
-                          const Eigen::Matrix<double, R2, C2> &v2) {
-  check_vector("dot_product", "v1", v1);
-  check_vector("dot_product", "v2", v2);
+template <typename T1, typename T2,
+          typename = require_all_eigen_vector_t<T1, T2>>
+inline auto dot_product(const T1 &v1, const T2 &v2) {
   check_matching_sizes("dot_product", "v1", v1, "v2", v2);
   return v1.dot(v2);
```

Comment:
These `check_vector` checks were redundant, as Eigen's `.dot()` is only defined for compile time vectors, so there is no way this could compile and check fail.